File: edgequake/sdks/typescript/src/resources/query.ts
/**
 * Query resource — execute RAG queries.
 *
 * @module resources/query
 * @see edgequake/crates/edgequake-api/src/handlers/query.rs
 */

import type { RequestOptions } from "../transport/types.js";
import type {
  QueryRequest,
  QueryResponse,
  QueryStreamEvent,
  StreamQueryRequest,
} from "../types/query.js";
import { Resource } from "./base.js";

export class QueryResource extends Resource {
  /** Execute a RAG query and get a complete response. */
  async execute(request: QueryRequest): Promise<QueryResponse> {
    return this._post("/api/v1/query", request);
  }

  /**
   * Execute a streaming RAG query.
   * Returns an async iterator of query stream events.
   *
   * WHY: The query stream endpoint returns raw text SSE chunks (not JSON),
   * so we cannot use _streamSSE which tries JSON.parse. Instead, we wrap
   * each raw text chunk as { chunk: data } to match QueryStreamEvent.
   */
  stream(
    request: StreamQueryRequest,
    signal?: AbortSignal,
  ): AsyncIterable<QueryStreamEvent> {
    const self = this;
    return {
      [Symbol.asyncIterator]() {
        return (async function* () {
          const options: RequestOptions = {
            method: "POST",
            path: "/api/v1/query/stream",
            body: request,
            signal,
          };
          for await (const data of self.transport.stream(options)) {
            // WHY: Stream data may be raw text OR JSON. Try JSON first.
            try {
              const parsed = JSON.parse(data) as QueryStreamEvent;
              yield parsed;
            } catch {
              // Raw text chunk — wrap as QueryStreamEvent
              yield { chunk: data };
            }
          }
        })();
      },
    };
  }
}

File: edgequake/sdks/python/edgequake/resources/query.py
"""Query resource — RAG query execution for EdgeQuake.

WHY: Maps to /api/v1/query endpoints. Supports both direct query execution
and streaming via SSE.
"""

from __future__ import annotations

from typing import Any, Literal

from edgequake._streaming import AsyncSSEStream, SSEStream
from edgequake.resources._base import AsyncResource, SyncResource
from edgequake.types.query import QueryResponse, QueryStreamEvent


class QueryResource(SyncResource):
    """Synchronous Query API."""

    def execute(
        self,
        query: str,
        *,
        mode: Literal["local", "global", "hybrid", "naive"] = "hybrid",
        top_k: int = 10,
        rerank: bool = False,
        provider: str | None = None,
        model: str | None = None,
        conversation_id: str | None = None,
    ) -> QueryResponse:
        """Execute a RAG query.

        POST /api/v1/query
        """
        body: dict[str, Any] = {
            "query": query,
            "mode": mode,
            "top_k": top_k,
            "rerank": rerank,
        }
        if provider:
            body["provider"] = provider
        if model:
            body["model"] = model
        if conversation_id:
            body["conversation_id"] = conversation_id
        return self._post("/api/v1/query", json=body, response_type=QueryResponse)

    def stream(
        self,
        query: str,
        *,
        mode: Literal["local", "global", "hybrid", "naive"] = "hybrid",
        top_k: int = 10,
    ) -> SSEStream[QueryStreamEvent]:
        """Execute a streaming RAG query via SSE.

        POST /api/v1/query/stream
        """
        body: dict[str, Any] = {
            "query": query,
            "mode": mode,
            "top_k": top_k,
        }
        response = self._transport.stream("POST", "/api/v1/query/stream", json=body)
        return SSEStream(response, QueryStreamEvent)


class AsyncQueryResource(AsyncResource):
    """Asynchronous Query API."""

    async def execute(
        self,
        query: str,
        *,
        mode: Literal["local", "global", "hybrid", "naive"] = "hybrid",
        top_k: int = 10,
        rerank: bool = False,
        provider: str | None = None,
        model: str | None = None,
        conversation_id: str | None = None,
    ) -> QueryResponse:
        body: dict[str, Any] = {
            "query": query,
            "mode": mode,
            "top_k": top_k,
            "rerank": rerank,
        }
        if provider:
            body["provider"] = provider
        if model:
            body["model"] = model
        if conversation_id:
            body["conversation_id"] = conversation_id
        return await self._post("/api/v1/query", json=body, response_type=QueryResponse)

    async def stream(
        self,
        query: str,
        *,
        mode: Literal["local", "global", "hybrid", "naive"] = "hybrid",
        top_k: int = 10,
    ) -> AsyncSSEStream[QueryStreamEvent]:
        body: dict[str, Any] = {
            "query": query,
            "mode": mode,
            "top_k": top_k,
        }
        response = await self._transport.stream(
            "POST", "/api/v1/query/stream", json=body
        )
        return AsyncSSEStream(response, QueryStreamEvent)
